File: lib/permissions/permission-service.ts
import { UserRole, Permission } from '@/types/common';

const rolePermissionsMap: Record<UserRole, Set<Permission>> = {
  [UserRole.SuperAdmin]: new Set([
    Permission.ManageUsers,
    Permission.ManageRoles,
    Permission.ViewAllLeads,
    Permission.ManageOwnLeads,
    Permission.AssignLeads,
    Permission.DeleteLead,
    Permission.ManageCustomers,
    Permission.ManageSociety,
    Permission.CreateServiceRequest,
    Permission.AssignTechnician,
    Permission.CompleteServiceRequest,
    Permission.ViewDashboardKpis,
    Permission.ViewFinancials,
  ]),
  [UserRole.Admin]: new Set([
    Permission.ViewAllLeads,
    Permission.ManageOwnLeads,
    Permission.AssignLeads,
    Permission.DeleteLead,
    Permission.ManageCustomers,
    Permission.ManageSociety,
    Permission.ManageUsers,
    Permission.CreateServiceRequest,
    Permission.AssignTechnician,
    Permission.CompleteServiceRequest,
    Permission.ViewDashboardKpis,
    Permission.ViewFinancials,
  ]),
  [UserRole.SalesManager]: new Set([
    Permission.ViewAllLeads,
    Permission.ManageOwnLeads,
    Permission.AssignLeads,
    Permission.ManageCustomers,
    Permission.ManageSociety,
    Permission.ViewDashboardKpis,
  ]),
  [UserRole.SalesExecutive]: new Set([
    Permission.ManageOwnLeads,
    Permission.ManageCustomers,
  ]),
  [UserRole.Technician]: new Set([
    Permission.CompleteServiceRequest,
  ]),
  [UserRole.Customer]: new Set([]),
};

export class PermissionService {
  static can(role: UserRole, permission: Permission): boolean {
    return rolePermissionsMap[role]?.has(permission) ?? false;
  }

  static permissionsFor(role: UserRole): Set<Permission> {
    return rolePermissionsMap[role] ?? new Set();
  }

  static getRoleLabel(role: UserRole): string {
    const labels: Record<UserRole, string> = {
      [UserRole.SuperAdmin]: 'Super Admin',
      [UserRole.Admin]: 'Admin',
      [UserRole.SalesManager]: 'Sales Manager',
      [UserRole.SalesExecutive]: 'Sales Executive',
      [UserRole.Technician]: 'Technician',
      [UserRole.Customer]: 'Customer',
    };
    return labels[role];
  }

  static getRoleColor(role: UserRole): string {
    const colors: Record<UserRole, string> = {
      [UserRole.SuperAdmin]: 'bg-purple-100 text-purple-800',
      [UserRole.Admin]: 'bg-red-100 text-red-800',
      [UserRole.SalesManager]: 'bg-blue-100 text-blue-800',
      [UserRole.SalesExecutive]: 'bg-green-100 text-green-800',
      [UserRole.Technician]: 'bg-yellow-100 text-yellow-800',
      [UserRole.Customer]: 'bg-gray-100 text-gray-800',
    };
    return colors[role];
  }
}

File: lib/features/role_access/services/permission_service.dart
import '../models/user_role.dart';

/// Single source of truth for "who can do what".
/// Add new permissions to [UserRole] enum's Permission list and wire them
/// here — never inline a role check in a screen.
class PermissionService {
  PermissionService._();

  static final Map<UserRole, Set<Permission>> _matrix = {
    UserRole.superAdmin: {
      Permission.manageUsers,
      Permission.manageRoles,
      Permission.viewAllLeads,
      Permission.manageOwnLeads,
      Permission.assignLeads,
      Permission.deleteLead,
      Permission.manageCustomers,
      Permission.manageSociety,
      Permission.createServiceRequest,
      Permission.assignTechnician,
      Permission.completeServiceRequest,
      Permission.viewDashboardKpis,
      Permission.viewFinancials,
    },
    UserRole.admin: {
      Permission.manageUsers,
      Permission.viewAllLeads,
      Permission.manageOwnLeads,
      Permission.assignLeads,
      Permission.deleteLead,
      Permission.manageCustomers,
      Permission.manageSociety,
      Permission.createServiceRequest,
      Permission.assignTechnician,
      Permission.completeServiceRequest,
      Permission.viewDashboardKpis,
      Permission.viewFinancials,
    },
    UserRole.salesManager: {
      Permission.viewAllLeads,
      Permission.manageOwnLeads,
      Permission.assignLeads,
      Permission.manageCustomers,
      Permission.manageSociety,
      Permission.viewDashboardKpis,
    },
    UserRole.salesExecutive: {
      Permission.manageOwnLeads,
      Permission.manageCustomers,
    },
    UserRole.technician: {
      Permission.completeServiceRequest,
    },
    UserRole.customer: {},
  };

  static bool can(UserRole role, Permission permission) {
    return _matrix[role]?.contains(permission) ?? false;
  }

  static Set<Permission> permissionsFor(UserRole role) => _matrix[role] ?? {};
}
